sycl/source/detail/graph_memory_pool.cpp

```diff
@@ -0,0 +1,189 @@
+    auto AllocOpt = tryReuseExistingAllocation(AlignedSize, AllocType,
+                                               AllocInfo.ReadOnly, DepNodes);
+
+    // If we got a value back then an allocation was available for reuse and we
+    // can just return that pointer
+    if (AllocOpt) {
+      return AllocOpt.value().Ptr;
+    }
+
+    // If no allocation could be reused, do a new virtual reservation
+    Adapter->call<sycl::errc::runtime,
+                  sycl::detail::UrApiKind::urVirtualMemReserve>(
+        CtxImpl->getHandleRef(), reinterpret_cast<void *>(StartPtr),
+        AlignedSize, &Alloc);
+
+    AllocInfo.Size = AlignedSize;
+    AllocInfo.Ptr = Alloc;
+
+    MAllocations[Alloc] = AllocInfo;
+    break;
+  }
+
+  default:
+    throw sycl::exception(sycl::make_error_code(errc::feature_not_supported),
+                          "Only device allocations are currently supported "
+                          "in graph allocation nodes!");
+    break;
+  }
+
+  return Alloc;
+}
+
+std::optional<graph_mem_pool::alloc_info>
+graph_mem_pool::tryReuseExistingAllocation(
+    size_t Size, usm::alloc AllocType, bool ReadOnly,
+    const std::vector<std::shared_ptr<node_impl>> &DepNodes) {
+
+  std::vector<alloc_info> CompatibleAllocs;
+  // Compatible allocs can only be as big as MFreeAllocations
+  CompatibleAllocs.reserve(MFreeAllocations.size());
+
+  // Loop over free allocation list, search for ones that are compatible for
+  // reuse Currently that means they have the same alloc kind, size and read
+  // only property.
+
+  for (auto &Ptr : MFreeAllocations) {
+    alloc_info &Info = MAllocations.at(Ptr);
+    if (Info.Kind == AllocType && Info.Size == Size &&
+        Info.ReadOnly == ReadOnly) {
+      // Store the alloc info since it is compatible
+      CompatibleAllocs.push_back(Info);
+    }
+  }
+
+  // Traverse graph back from each DepNode to try and find any of the suitable
+  // free nodes. We do this in a breadth-first approach because we want to find
+  // the shortest path to a reusable allocation.
+
+  std::queue<std::weak_ptr<node_impl>> NodesToCheck;
+
+  // Add all the dependent nodes to the queue, they will be popped first
+  for (auto &Dep : DepNodes) {
+    NodesToCheck.push(Dep);
+  }
+
+  std::optional<alloc_info> AllocInfo = {};
+
+  // Called when traversing over nodes to check if the current node
+  auto CheckNodeEqual =
+      [&CompatibleAllocs,
+       &AllocInfo](const std::shared_ptr<node_impl> &CurrentNode) -> bool {
+    for (auto &Alloc : CompatibleAllocs) {
+      const auto &AllocFreeNode = Alloc.LastFreeNode;
+      // Compare control blocks without having to lock AllocFreeNode to check
+      // for node equality
+      if (!CurrentNode.owner_before(AllocFreeNode) &&
+          !AllocFreeNode.owner_before(CurrentNode)) {
+        Alloc.LastFreeNode.reset();
+        AllocInfo = Alloc;
+        return true;
+      }
+    }
+    return false;
+  };
+
+  while (!NodesToCheck.empty()) {
```

Comment:
Can we avoid the graph traversal when `CompatibleAllocs` is empty?

Reply:
I've added an early return for this case